Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm

```diff
@@ -41,46 +47,162 @@
 {
-    ASSERT(m_streamType != NetworkTransportStreamType::OutgoingUnidirectional);
+    if (m_streamState == NetworkTransportStreamState::ReadClosed) {
+        RELEASE_ASSERT_NOT_REACHED();
+        return;
+    }
     nw_connection_receive(m_connection.get(), 0, std::numeric_limits<uint32_t>::max(), makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t content, nw_content_context_t, bool withFin, nw_error_t error) {
         RefPtr strongThis = weakThis.get();
         if (!strongThis)
             return;
-        if (error)
-            return; // FIXME: Pipe this error to JS.
-        if (!content && withFin)
+        RefPtr session = strongThis->m_session.get();
+        if (!session)
+            return;
+        if (error) {
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                session->streamReceiveBytes(strongThis->m_identifier, { }, false, WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
             return;
+        }
+
+        ASSERT(content || withFin);
 
         // FIXME: Not only is this an unnecessary string copy, but it's also something that should probably be in WTF or FragmentedSharedBuffer.
         auto vectorFromData = [](dispatch_data_t content) {
-            ASSERT(content);
             Vector<uint8_t> request;
-            dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
-                request.append(buffer);
-                return true;
-            });
+            if (content) {
+                dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
+                    request.append(buffer);
+                    return true;
+                });
+            }
             return request;
         };
 
-        if (RefPtr session = strongThis->m_session.get())
-            session->streamReceiveBytes(strongThis->m_identifier, vectorFromData(content).span(), withFin);
-        if (!withFin)
+        session->streamReceiveBytes(strongThis->m_identifier, vectorFromData(content).span(), withFin, std::nullopt);
+
+        if (withFin) {
+            switch (strongThis->m_streamState) {
+            case NetworkTransportStreamState::Ready:
+                strongThis->m_streamState = NetworkTransportStreamState::ReadClosed;
+                break;
+            case NetworkTransportStreamState::WriteClosed:
+                strongThis->cancelReceive(std::nullopt);
+                break;
+            case NetworkTransportStreamState::ReadClosed:
+                RELEASE_ASSERT_NOT_REACHED();
+            }
+        } else
             strongThis->receiveLoop();
     }).get());
 }
+
+void NetworkTransportStream::setErrorCodeForStream(std::optional<WebCore::WebTransportStreamErrorCode> errorCode)
+{
+    if (!errorCode)
+        return;
+
+    // FIXME: Implement once rdar://141886375 is available in OS builds.
```

Comment:
We could probably do this before then with optional soft linking.

Reply:
I'll add this in a separate PR.